File: akkademy-scala-db/src/main/scala/com/akkademy/AkkademyDb.scala
package com.akkademy

import akka.actor.Actor
import scala.collection.mutable.HashMap
import akka.event.Logging
import com.akkademy.message.SetRequest
import akka.actor.Status
import com.akkademy.message.GetRequest
import com.akkademy.message.KeyNotFoundException
import akka.actor.ActorSystem
import akka.actor.Props
import com.akkademy.message.SetIfNoExist
import com.akkademy.message.KeyExistException
import com.akkademy.message.DeleteRequest

class AkkademyDb extends Actor{
  
  val map = new HashMap[String, Object]
  val log = Logging(context.system, this)
  
  override def receive = {
    case SetRequest(key, value) => {
      log.info("Received SetRequest - key: {} value: {}", key, value)
      map.put(key, value)
      sender ! Status.Success
    }
    case GetRequest(key) => {
      log.info("Received GetRequest - key: {} ", key)
      val response: Option[Object] = map.get(key)
      response match{
        case Some(x) => sender() ! x
        case None =>  sender()  ! Status.Failure(new KeyNotFoundException(key))
      }
    }
    case SetIfNoExist(key, value) => {
      log.info("Received SetIfNoExist - key: {} value: {}", key, value)
      val response: Option[Object] = map.get(key)
      response match {
        case Some(x) => sender() ! Status.Failure(new KeyExistException(key))
        case None => {
          map.put(key, value)
          sender() ! Status.Success
        }
      }
    }
    case DeleteRequest(key) => {
      log.info("Received DeleteRequest - key: {}", key)
      val response: Option[Object] = map.remove(key)
      response match{
        case Some(x) => sender() ! Status.Success
        case None => sender() ! Status.Failure(new KeyNotFoundException(key))
      }
    }
    case o => Status.Failure(new ClassNotFoundException)
  }
}

object Main extends App{
  val system = ActorSystem("akkademy")
  system.actorOf(Props[AkkademyDb], name = "akkademy-db")
}

File: akkademy-db-java/src/main/java/com/akkademy/AkkademyDb.java
package com.akkademy;

import java.util.HashMap;
import java.util.Map;

import akka.actor.AbstractActor;
import akka.actor.Status;
import akka.event.Logging;
import akka.event.LoggingAdapter;
import akka.japi.pf.ReceiveBuilder;

import com.akkademy.messages.DeleteRequest;
import com.akkademy.messages.GetRequest;
import com.akkademy.messages.KeyExistException;
import com.akkademy.messages.KeyNotFoundException;
import com.akkademy.messages.SetIfNotExist;
import com.akkademy.messages.SetRequest;

public class AkkademyDb extends AbstractActor {
	
	protected final LoggingAdapter log = Logging.getLogger(context().system(), this);
	protected final Map<String, Object> map = new HashMap<>();
	
	private AkkademyDb(){
		receive(ReceiveBuilder
			.match(SetRequest.class, message -> {
				log.info("Received Set request - key: {} value: {}",message.key, message.value);
				map.put(message.key, message.value);
				sender().tell(new Status.Success(message.key), self());
			})
			.match(GetRequest.class, message -> {
				log.info("Received Get request - key: {} ",message.key);
				Object value = map.get(message.key);
				Object response = (value != null)
						? value 
						: new Status.Failure(new KeyNotFoundException(message.key));
				sender().tell(response, self());
			})
			.match(SetIfNotExist.class, message -> {
				log.info("Received SetIfExist request - key: {} value: {}",message.key, message.value);
				Object value = map.get(message.key);
				Object response;
				if(value != null){
					response = new Status.Failure(new KeyExistException(message.key)); 
				}else{
					map.put(message.key, message.value);
					response = new Status.Success(message.key);
				}
				sender().tell(response, self());
			})
			.match(DeleteRequest.class, message -> {
				log.info("Received Delete request - key: {}",message.key);
				Object value = map.remove(message.key);
				Object response = (value != null)
						? new Status.Success(message.key) 
						: new Status.Failure(new KeyNotFoundException(message.key));
				sender().tell(response, self());
			})
			.matchAny(o -> sender().tell(new Status.Failure(new ClassNotFoundException()), self()))
			.build());
	}
}
